### PR TITLE
[RLlib] Support terminated and truncated in ExternalMultiAgentEnv 

### LINKED ISSUE #35189: [RLlib] `ExternalMultiAgentEnv` yields error on `log_returns` with `multiagent_done_dict`
### What happened + What you expected to happen

When using `log_returns` of the `ExternalMultiAgentEnv` with a provided `multiagent_done_dict`, an error is raised stating that `_ExternalEnvEpisode` has no `cur_done_dict` attribute.

Here is the full output:
```
Ray version: 2.3.1
Traceback (most recent call last):
  File "./07_external_multienv_github_issue.py", line 6, in <module>
    env.log_returns(eid, { "policy": 1.0 }, {}, { "policy": True})
  File "./venv/lib/python3.10/site-packages/ray/rllib/env/external_multi_agent_env.py", line 154, in log_returns
    episode.cur_done_dict[agent] = done
AttributeError: '_ExternalEnvEpisode' object has no attribute 'cur_done_dict'. Did you mean: 'cur_info_dict'?
```

The problem seems to be [here](https://github.com/ray-project/ray/blob/cc3fa331b8599c86da0078bce93fe6d84dba1f87/rllib/env/external_multi_agent_env.py#L154) and/or [here](https://github.com/ray-project/ray/blob/cc3fa331b8599c86da0078bce93fe6d84dba1f87/rllib/env/external_env.py#L260-L266).

### Versions / Dependencies

Ray version: 2.3.1
Python: 3.10
Windows 10 using WSL with Ubuntu 22.04.2 LTS

### Reproduction script

Really simple example without `PolicyServerInput` & `PolicyClient`. The snippet outputs the formerly mentioned error.

```python
from ray.rllib.env.external_multi_agent_env import ExternalMultiAgentEnv
from gymnasium.spaces import Discrete, Box
import ray

print("Ray version:", ray.__version__)

env = ExternalMultiAgentEnv(Discrete(2), Box(float("-inf"), float("inf"), (1,)))
eid = env.start_episode(training_enabled=True)
env.log_returns(eid, { "policy": 1.0 }, {}, { "policy": True})
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by Paolo-Fermin**
## Why are these changes needed?

The ExternalMultiAgentEnv class accepts a `multiagent_done_dict` as a part of the `log_returns` function. However, given the recent change to the gym API to gymnasium, dones have been split into terminated and truncated booleans. This PR updates the ExternalMultiAgentEnv api to accept `multiagent_terminated_dict` as well as `multiagent_truncated_dict`, while retaining the old `multiagent_done_dict` for backwards compability. In the case that the user provides the deprecated `multiagent_done_dict`, assume that they want to set both terminated and truncated simultaneously. 

## Related issue number

Closes #35189 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by dogganon** [comment]
# It's been a year and THIS PR STILL HAS NOT BEEN MERGED??

@ericl @sven1977 That being said, y'all NEED TO SERIOUSLY take a look at the related https://github.com/ray-project/ray/issues/35189 to see how severe the bug IS!! **AND MERGE THIS PR ASAP!!!!** [`multiagent_done_dict`](https://github.com/ray-project/ray/blob/d5fa9a04ed841ea845887f43e06a0d2a81216c2d/rllib/env/external_multi_agent_env.py#L145C25-L145C38) in `ray.rllib.env.ExternalMultiAgentEnv.log_returns` IS BASICALLY UNUSABLE DUE TO `ray.rllib.env.external_env._ExternalEnvEpisode.cur_done_dict` BEING NOT DEFINED! THIS COULD HAVE BEEN EASILY PREVENTED IF THERE WERE AT LEAST SOME UNIT TESTS IN PLACE! MIND BLOWING!!

# SO TLDR, MERGE THIS PR ASAP AND ADD UNIT TESTS FOR `ray.rllib.env.ExternalMultiAgentEnv`!! 🤯

@sven1977
@gjoliver
@avnishn
@ArturNiederfahrenhorst
@smorad
@maxpumperla
@kouroshHakha
@krfricke
